GFDL->main PR regression and bug fixes

marshallward opened the pull request:
Several fixes to the candidate PR to main:
* Fixed bugs due to introduction of MEKE biharmonic friction
* Restored performance due to new visc-limit flag in MEKE FrictWork.  This is now applied separately, and only if `EY24_EBT_BR` is enabled.
* Restored performance due to multiple damped MEKE diagnostics.  Nearly every calculation is now conditional.
* Bugfix to an internal tide control struct declaration

There is also a commit to restore answers related to MEKE backscatter, but this is still under some discussion.  I'm leaving it in for now, but we may change or remove this one if NCAR approves.

This needs review from one of the following:
* @wendazhang33
* @elizabethYankovsky

Further testing will eventually be needed from @alperaltuntas and @jiandewang but this can wait until Wenda and/or Elizabeth have reviewed the modificatins.

marshallward commented:
The regression is due to the `visc_lim_*` diagnostics, which are now conditionally registered if `CS%EY24_EBT_BS` is enabled.

Although this is a regression into [NOAA-GFDL:gfdl-to-main-2024-11-27](https://github.com/NOAA-GFDL/MOM6/tree/gfdl-to-main-2024-11-27) it should not cause a regression of [NOAA-GFDL:gfdl-to-main-2024-11-27](https://github.com/NOAA-GFDL/MOM6/tree/gfdl-to-main-2024-11-27) into `main`.

Wendazhang33 commented:
Thanks for the PR. I have reviewed the changes and made a few comments. Please let me know if you have any questions.
________________________________
From: Marshall Ward ***@***.***>
Sent: Thursday, December 19, 2024 12:42 PM
To: NOAA-GFDL/MOM6 ***@***.***>
Cc: Wenda Zhang ***@***.***>; Mention ***@***.***>
Subject: [NOAA-GFDL/MOM6] GFDL->main PR regression and bug fixes (PR #782)


Several fixes to the candidate PR to main:

  *   Fixed bugs due to introduction of MEKE biharmonic friction
  *   Restored performance due to new visc-limit flag in MEKE FrictWork. This is now applied separately, and only if EY24_EBT_BR is enabled.
  *   Restored performance due to multiple damped MEKE diagnostics. Nearly every calculation is now conditional.
  *   Bugfix to an internal tide control struct declaration

There is also a commit to restore answers related to MEKE backscatter, but this is still under some discussion. I'm leaving it in for now, but we may change or remove this one if NCAR approves.

This needs review from one of the following:

  *   @Wendazhang33<https://github.com/Wendazhang33>
  *   @ElizabethYankovsky<https://github.com/ElizabethYankovsky>

Further testing will eventually be needed from @alperaltuntas<https://github.com/alperaltuntas> and @jiandewang<https://github.com/jiandewang> but this can wait until Wenda and/or Elizabeth have reviewed the modificatins.

________________________________
You can view, comment on, or merge this pull request online at:

  https://github.com/NOAA-GFDL/MOM6/pull/782

Commit Summary

  *   e59e2d6<https://github.com/NOAA-GFDL/MOM6/pull/782/commits/e59e2d694d487f2d4cec2aa93d4e18e101b17716> MEKE: Split src and src_GM compute loops
  *   0a7ca20<https://github.com/NOAA-GFDL/MOM6/pull/782/commits/0a7ca20b6a97b25a4e6618e6d13d936b47564072> MEKE: Conditionally compute biharmonic FrCoeff
  *   423e388<https://github.com/NOAA-GFDL/MOM6/pull/782/commits/423e388f9caada870fea87c006f1e80dcf635e43> MEKE: Revert Backscatter expressions
  *   7e32c83<https://github.com/NOAA-GFDL/MOM6/pull/782/commits/7e32c83d62c7564f457bfb48e475db3ddceacc96> MEKE: Move flag test outside of FrictWork loop
  *   f20a98c<https://github.com/NOAA-GFDL/MOM6/pull/782/commits/f20a98c420d83a7380e021979a814de98c87a33a> MEKE: Move damping diagnostics outside MEKE loop
  *   5f4bbc2<https://github.com/NOAA-GFDL/MOM6/pull/782/commits/5f4bbc25efe891af8e89306bf6201409a71f88c4> Diffusivity: Revert int_tide_CS to pointer

File Changes

(3 files<https://github.com/NOAA-GFDL/MOM6/pull/782/files>)

  *   M src/parameterizations/lateral/MOM_MEKE.F90<https://github.com/NOAA-GFDL/MOM6/pull/782/files#diff-2bf6c52474ad3142c4902b0f94ed67365c19e0bc7a6a61c4dcbc4e453330d258> (250)
  *   M src/parameterizations/lateral/MOM_hor_visc.F90<https://github.com/NOAA-GFDL/MOM6/pull/782/files#diff-66ca61ccaca43a9217e857b349d1ffc3fdfcfa65b04995b6e62b0d66d1e1a16d> (218)
  *   M src/parameterizations/vertical/MOM_set_diffusivity.F90<https://github.com/NOAA-GFDL/MOM6/pull/782/files#diff-f0477334759277205ea346ab76bda77c93199f160413c86e25d59a8aa7088c00> (2)

Patch Links:

  *   https://github.com/NOAA-GFDL/MOM6/pull/782.patch
  *   https://github.com/NOAA-GFDL/MOM6/pull/782.diff

—
Reply to this email directly, view it on GitHub<https://github.com/NOAA-GFDL/MOM6/pull/782>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AKBBAQFWVTKXSRBKQCHBM432GMASFAVCNFSM6AAAAABT5P6X6WVHI2DSMVQWIX3LMV43ASLTON2WKOZSG42TCMBVG4YDGOI>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

marshallward commented on src/parameterizations/lateral/MOM_MEKE.F90:
```diff
@@ -487,32 +535,74 @@ subroutine step_forward_MEKE(MEKE, h, SN_u, SN_v, visc, dt, G, GV, US, CS, hu, h
+      MEKE%MEKE(i,j) =  MEKE%MEKE(i,j) / (1. + sdt_damp * damp_rate(i,j))
+    enddo ; enddo
+
+    if (any_damping_diags) then
```
Thanks for catching that, it's fixed.

marshallward commented on src/parameterizations/lateral/MOM_MEKE.F90:
```diff
@@ -675,30 +771,80 @@ subroutine step_forward_MEKE(MEKE, h, SN_u, SN_v, visc, dt, G, GV, US, CS, hu, h
+          !$OMP parallel do default(shared)
+          do j=js,je ; do i=is,ie
+            src_btm_drag(i,j) = -MEKE_current(i,j) * ( &
+                damp_step * (damp_rate(i,j) * damping(i,j)) &
```
Well, bug or not, I certainly did a bad job transcribing it :grimacing:.  I've restore the expression.

Do you think this should be fixed now?  Or better to do in a later PR?

Wendazhang33 replied:
This computation has probably been used by many people. I think maybe we should keep it in the interest of reproducibility. If we are sure there is a bug later, we can submit a PR adding a bug flag. What do you think?

marshallward replied:
The diagnostic will be new to `main`, so it depends if anyone at GFDL needs reproducibility.  Who can we ask about this?

marshallward replied:
@Wendazhang33 and I agreed that this should be OK in its current form.

marshallward commented on src/parameterizations/lateral/MOM_MEKE.F90:
```diff
@@ -412,22 +415,64 @@ subroutine step_forward_MEKE(MEKE, h, SN_u, SN_v, visc, dt, G, GV, US, CS, hu, h
+        src(i,j) = src(i,j) - bh_coeff * I_mass(i,j) * MEKE%mom_src_bh(i,j)
       enddo ; enddo
+
+      if (CS%id_src_mom_lp > 0) then
```
Would one ever register `src_mom_lp` if `MEKE_FrCoeff` were unset?  Maybe it's better to address this possibility at initialization, rather than here.

Wendazhang33 replied:
Agree. We should only register src_mom_lp when MEKE_FrCoeff is set.

marshallward replied:
I've moved the following diagnostics into conditional blocks:

* `id_src_mom_lp`
* `id_src_mom_bh`
* `id_src_GM`
* `id_src_mom_K4`

marshallward commented on src/parameterizations/lateral/MOM_MEKE.F90:
```diff
@@ -412,22 +415,64 @@ subroutine step_forward_MEKE(MEKE, h, SN_u, SN_v, visc, dt, G, GV, US, CS, hu, h
+        CS%id_src_mom_K4 > 0 &
+    ])
+
     if (allocated(MEKE%mom_src)) then
```
Agreed, this is a better criterion, thanks.  I've replaced the condition.

marshallward commented on src/parameterizations/lateral/MOM_hor_visc.F90:
```diff
@@ -2159,26 +2165,58 @@ subroutine horizontal_viscosity(u, v, h, uh, vh, diffu, diffv, MEKE, VarMix, G,
-          MEKE%mom_src(i,j) = MEKE%mom_src(i,j) + (FrictWork(i,j,k) - RoScl*FrictWork_bh(i,j,k))
-          MEKE%mom_src_bh(i,j) = MEKE%mom_src_bh(i,j) + &
-                                 (FrictWork_bh(i,j,k) - RoScl*FrictWork_bh(i,j,k))
+          MEKE%mom_src(i,j) = MEKE%mom_src(i,j) + GV%H_to_RZ * ( &
```
Thanks for catching this error.

After discussing with NCAR via @alperaltuntas it appears unnecessary to retain answers (bug included) when `MEKE_BACKSCAT` is set, so I have removed the commit which reverted these expressions.

`MEKE%mom_src(i,j) + (FrictWork(i,j,k) - RoScl*FrictWork_bh(i,j,k))` is now always used, which is mathematically correct for both true and false `FrictWork_bug`.

marshallward commented:
I've made several adjustments based on suggestions by @Wendazhang33

* The reverted `mom_src` update for `MEKE_BACKSCAT_RO_C` was un-reverted.  Although this expression was required to produce the original bit-reproducible solutions, it was also wrong for two reasons:
  
  * Code following the `endif` should have been in an `else`-block, causing terms to be double-computed.  (I had originally prepared a bug flag to replicate this effect, but that was also removed.)
  
  * The expression would have only been correct for `FrictWork_bug = True`.
  
  The new expression is both concise and correct for both values of `FrictWork_bug`, and NCAR supports the existing answer-changing expression, so I have removed this commit.

* Some flow control is more accurately based on `FrCoeff` rather than allocation state.

* Diagnostics which depend on `FrCoeff` are not registered if `FrCoeff` is unset.

* Fixed two bugs caught by Wenda

I believe this is now ready for review, and merge into the PR to main.

Hallberg-NOAA commented on src/parameterizations/lateral/MOM_hor_visc.F90:
```diff
@@ -1999,19 +1995,21 @@ subroutine horizontal_viscosity(u, v, h, uh, vh, diffu, diffv, MEKE, VarMix, G,
+      if (CS%FrictWork_bug) then
         ! Diagnose   bhstr_xx*d_x u - bhstr_yy*d_y v + bhstr_xy*(d_y u + d_x v)
         ! This is the old formulation that includes energy diffusion !cyc
+        do j=js,je ; do i=is,ie
```
The following new expression for FrictWork_bh is lacking the parentheses that would be needed to give rotational symmetry when FMAs are enabled.  Compare this with the analogous expression for FrictWork at lines 1952-1966 to see where the parentheses would be needed.

marshallward replied:
Thanks Bob, I added parentheses which are are hopefully correct and should preserve FMA rotation symmetry.

Hallberg-NOAA replied:
Thank you for adding these parentheses.  I have examined them carefully and they should now preserve FMA symmetry.

Hallberg-NOAA reviewed (commented):
I have examined all of these changes, and I have checked out the updated code and used it to run the MOM6-examples regression suite.  The change in diagnostics in the TC testing is properly explained.  I am satisfied that this is correct and that it is likely to address the issues that were identified with the 2024-11-27 GFDL-to-main (https://github.com/mom-ocean/MOM6/pull/1647) pull request.

marshallward commented:
Gaea regression: https://gitlab.gfdl.noaa.gov/ogrp/mom6ci/MOM6/-/pipelines/25932 :heavy_check_mark: :yellow_circle:

This passes our own tests, I will merge this into the PR for consortium review.
